packages/world/src/systems/TransferSystem.sol

```diff
@@ -4,142 +4,146 @@ pragma solidity >=0.8.24;
+
+import { transferInventoryEntity, transferInventoryNonEntity } from "../utils/InventoryUtils.sol";
 import { notify, TransferNotifData } from "../utils/NotifUtils.sol";
-import { TransferLib } from "./libraries/TransferLib.sol";
+import { callProgramOrRevert } from "../utils/callProgram.sol";
+import { getForceField } from "../utils/ForceFieldUtils.sol";
+import { updateMachineEnergy, transferEnergyToPool } from "../utils/EnergyUtils.sol";
+
+import { ProgramOnTransferData } from "../Types.sol";
 import { EntityId } from "../EntityId.sol";
 import { ObjectTypeId } from "../ObjectTypeId.sol";
-import { callProgramOrRevert } from "../utils/callProgram.sol";
+import { ObjectTypes } from "../ObjectTypes.sol";
+import { ObjectAmount } from "../ObjectTypeLib.sol";
+import { Vec3 } from "../Vec3.sol";
+import { SMART_CHEST_ENERGY_COST } from "../Constants.sol";
 
 contract TransferSystem is System {
-  function requireAllowed(
-    uint256 machineEnergyLevel,
-    bool isDeposit,
-    EntityId playerEntityId,
-    EntityId chestEntityId,
-    TransferData memory transferData,
-    bytes calldata extraData
-  ) internal {
-    if (machineEnergyLevel > 0) {
-      // Forward any ether sent with the transaction to the hook
-      // Don't safe call here as we want to revert if the program doesn't allow the transfer
-      bytes memory onTransferCall = abi.encodeCall(
-        IChestProgram.onTransfer,
-        (
-          ProgramOnTransferData({
-            targetEntityId: chestEntityId,
-            callerEntityId: playerEntityId,
-            isDeposit: isDeposit,
-            transferData: transferData,
-            extraData: extraData
-          })
-        )
-      );
-      callProgramOrRevert(chestEntityId.getProgram(), onTransferCall);
-    }
-  }
-
   function transfer(
-    EntityId chestEntityId,
-    bool isDeposit,
+    EntityId callerEntityId,
+    EntityId fromEntityId,
+    EntityId toEntityId,
     ObjectTypeId transferObjectTypeId,
     uint16 numToTransfer,
     bytes calldata extraData
   ) public payable {
-    TransferCommonContext memory ctx = TransferLib.transferCommon(_msgSender(), chestEntityId, isDeposit);
-    transferInventoryNonEntity(
-      ctx.isDeposit ? ctx.playerEntityId : ctx.chestEntityId,
-      ctx.isDeposit ? ctx.chestEntityId : ctx.playerEntityId,
-      ctx.dstObjectTypeId,
-      transferObjectTypeId,
-      numToTransfer
-    );
+    callerEntityId.activate();
 
-    notify(
-      ctx.playerEntityId,
-      TransferNotifData({
-        transferEntityId: ctx.isDeposit ? ctx.chestEntityId : ctx.playerEntityId,
-        transferCoord: ctx.chestCoord,
-        transferObjectTypeId: transferObjectTypeId,
-        transferAmount: numToTransfer
-      })
-    );
+    if (callerEntityId != fromEntityId) {
+      callerEntityId.requireConnected(fromEntityId);
+    }
+    fromEntityId.requireConnected(toEntityId);
+
+    transferEnergyToPool(callerEntityId, SMART_CHEST_ENERGY_COST);
```

Comment:
This energy cost would still run even if the chest doesnt have a force field right? or if the force field is out of charge, but you should still be able to take/withdraw

Reply:
yes, should there be a cost for transfers? should it be drained from the forcefield instead?

Reply:
I was thinking initially no, but then maybe yes to be consistent with chest callers

1) player transfer items --> take energy from player
2) chest transfer items --> take energy from force field

I think 1) could technically drain the force field by just going in and out if it took from the force field, so better if it takes from the player. for 2) the force field gets to control the chests inside

Reply:
so then the current code is actually good as it is